### ISSUE TITLE
kube-scheduler binary bump to 1.16 make user-scheduler in 1.15 k8s fail

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/zero-to-jupyterhub-k8s#1773]
#1699 introduced an breaking bug using this Helm chart on Kubernetes clusters versioned 1.15 and lower. The `user-scheduler` pods responsible for scheduling the k8s pods representing the users in a smart manner is now failing to schedule them to the k8s nodes with the following error message.

```
I0908 10:05:43.912366       1 reflector.go:158] Listing and watching *v1beta1.CSINode from k8s.io/client-go/informers/factory.go:134
E0908 10:05:43.914298       1 reflector.go:123] k8s.io/client-go/informers/factory.go:134: Failed to list *v1beta1.CSINode: csinodes.storage.k8s.io is forbidden: User "system:serviceaccount:jupyterhub:user-scheduler" cannot list resource "csinodes" in API group "storage.k8s.io" at the cluster scope
```

The error describes that the user-scheduler pod isn't allowed to ask the k8s api-server about `storage.k8s.io/v1beta/csinodes` resources.

To understand why this is, we need to understand what grants the user-scheduler pod permission to potentially do something like this in the first place.

1. The user-scheduler pod is configured to use a k8s ServiceAccount named `user-scheduler`.
2. The k8s ServiceAccount `user-scheduler` is coupled to have the Cluster permissions of...
   - the ClusterRole `system:kube-scheduler` through the ClusterRoleBinding `jupyterhub-user-scheduler-base`
   - the ClusterRole `jupyterhub-user-scheduler-complementary` through the ClusterRoleBinding `jupyterhub-user-scheduler-complementary`.
3. The permissions defined in ClusterRole `system:kube-scheduler` depends on the k8s version, its not part of this Helm chart. And, k8s 1.15 doesn't include the permission to inspect CSINode resources, while k8s 1.16's version of the `system:kube-scheduler` ClusterRole does.

Due to this, the solution as I see it, is to add the permission to inspect CSINode resources in the ClusterRole `jupyterhub-user-scheduler-complementary`. This is a one-liner fix right in this section, by adding `csinodes` next to `storageclasses`.

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/cd1eff7a5f093453de4d5868d1c4148044c0db23/jupyterhub/templates/scheduling/user-scheduler/rbac.yaml#L39-L42

---

/cc: @deepitpro and @ArijitBhattacharyya11 who have also seen the consequences of this issue - perhaps you could create a PR to fix this also? =) Following its merge, a new developer release version will be published herehttps://jupyterhub.github.io/helm-chart/ which we can upgrade to.

### Related

Closes #1773.

**Comment by consideRatio** [comment]
# Deep dive in the permissions required for the kube-scheduler binary

It seems like the `kube-scheduler` binary has been using two ClusterRoles that have evolved over time, the `system:kube-scheduler` binary, and the `system:volume-scheduler` binary. The `system:kube-scheduler` ClusterRole doesn't add permissions to inspect a `storageclass` for example, but we have found that we need to add that for our `user-scheduler` which is a configured `kube-scheduler` binary to work.

## Conclusions
We should provide complementary permissions to our user-scheduler binary, both from `system:kube-scheduler` and `system:volume-scheduler`. For example, if we assume this Helm chart could end up running in Kubernetes 1.15, and we use the `kube-scheduler` binary published associated with k8s 1.16, then we should add the permissions that a `kube-scheduler` binary expects in `system:kube-scheduler` and `system:volume-scheduler` that wasn't part of k8s 1.15 but is part of k8s 1.16's version of these ClusterRoles. Due to this, I've made an inspection of the changes over time to the relevant ClusterRole's.

## Action points
- [ ] Reference `system:volume-scheduler` and remove the permissions added to `jupyterhub-user-scheduler-complementary` that was added there because we didn't understand that we should also reference the `system:volume-scheduler` ClusterRole for our user-scheduler's ServiceAccount.
- [ ] Ensure we have all permissions up to 1.19 versions of kube-scheduler for our `user-scheduler`, among them are the `csinodes` for example.

## Relevant links

- The [system:kube-scheduler ClusterRole](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#core-component-roles)
- The [kube-scheduler configuration new for 1.19 (beta)](https://kubernetes.io/docs/reference/scheduling/config/)
- The [kube-scheduler binary reference](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/)

## ClusterRole `system:volume-scheduler`

No changes between [1.12](https://github.com/kubernetes/kubernetes/blob/v1.12.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L1114-L1142) and [1.19](https://github.com/kubernetes/kubernetes/blob/v1.19.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L1212-L1240).

## ClusterRole `system:kube-scheduler`

- [1.19.0 definitions](https://github.com/kubernetes/kubernetes/blob/v1.19.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L516-L558) and [output](https://github.com/kubernetes/kubernetes/blob/v1.19.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L696-L829)
- [1.18.0 definitions](https://github.com/kubernetes/kubernetes/blob/v1.18.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L517-L553) and [output](https://github.com/kubernetes/kubernetes/blob/v1.18.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L696-L829)
- [1.17.0 definitions](https://github.com/kubernetes/kubernetes/blob/v1.17.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L476-L511) and [output](https://github.com/kubernetes/kubernetes/blob/v1.17.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L624-L757)
- [1.16.0 definitions](https://github.com/kubernetes/kubernetes/blob/v1.16.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L480-L530) and [output](https://github.com/kubernetes/kubernetes/blob/v1.16.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L710-L830)
- [1.15.0 definitions](https://github.com/kubernetes/kubernetes/blob/v1.15.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L420-L445) and [output](https://github.com/kubernetes/kubernetes/blob/v1.15.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L692-L803)
- [1.14.0 definitions](https://github.com/kubernetes/kubernetes/blob/v1.14.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L419-L444) and [output](https://github.com/kubernetes/kubernetes/blob/v1.14.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L711-L822)
- [1.13.0 definitions](https://github.com/kubernetes/kubernetes/blob/v1.13.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L415-L437) and [output](https://github.com/kubernetes/kubernetes/blob/v1.13.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L699-L798)
- [1.12.0 definitions](https://github.com/kubernetes/kubernetes/blob/v1.12.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L414-L436) and [output](https://github.com/kubernetes/kubernetes/blob/v1.12.0/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml#L697-L796)

## Summary of changes

- 1.12 initial ClusterRole considered
- 1.13 no change
- 1.14 no change
- 1.16 made a change
- 1.17 made a change
- 1.18 no change
- 1.19 no change

### The 1.16 change

```diff
@@ -1,6 +1,7 @@
   rules:
   - apiGroups:
     - ""
+    - events.k8s.io
     resources:
     - events
     verbs:
@@ -98,3 +99,23 @@
     - get
     - list
     - watch
+  - apiGroups:
+    - authentication.k8s.io
+    resources:
+    - tokenreviews
+    verbs:
+    - create
+  - apiGroups:
+    - authorization.k8s.io
+    resources:
+    - subjectaccessreviews
+    verbs:
+    - create
+  - apiGroups:
+    - storage.k8s.io
+    resources:
+    - csinodes
+    verbs:
+    - get
+    - list
+    - watch
```

### The 1.17 change

```diff
@@ -8,6 +8,21 @@
     - create
     - patch
     - update
+  - apiGroups:
+    - coordination.k8s.io
+    resources:
+    - leases
+    verbs:
+    - create
+  - apiGroups:
+    - coordination.k8s.io
+    resourceNames:
+    - kube-scheduler
+    resources:
+    - leases
+    verbs:
+    - get
+    - update
   - apiGroups:
     - ""
     resources:
@@ -21,9 +36,7 @@
     resources:
     - endpoints
     verbs:
-    - delete
     - get
-    - patch
     - update
   - apiGroups:
     - ""
```

**Comment by consideRatio** [comment]
## Report following struggles

### Fix for kube-scheduler 1.16
`kube-scheduler` 1.16 can be used on k8s 1.15+ like described in the original post by adding permissions to inspect CSINode resources.

### Fix for kube-scheduler 1.17+

These `kube-scheduler` binaries will try to inspect CSINode resources, but when they fail they will not start their scheduling loop.

```
Failed to watch *v1.CSINode: failed to list *v1.CSINode: the server could not find the requested resource
```

### Modern kube-scheduler configuration
In k8s 1.17 (v1alpha1), k8s 1.18 (v1alpha2), and k8s 1.19 (v1beta1), the KubeSchedulerConfiguration format is available. This replaces some command line flags and configuration in another format previously used to specify how the kube-scheduler should act. [This reference](https://pkg.go.dev/k8s.io/kube-scheduler@v0.19.0/config/v1beta1?tab=doc#KubeSchedulerConfiguration) can be used to inspect details of the format.

In that format, our policy to schedule on the most used node end up a lot simpler.

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta1
kind: KubeSchedulerConfiguration
profiles:
  - schedulerName: {{ .Release.Name }}-user-scheduler
    plugins:
      score:
        disabled:
          - name: NodeResourcesLeastAllocated
          - name: NodeResourcesBalancedAllocation
        enabled:
          - name: NodeResourcesMostAllocated
```
